project5/3700kvstore

```diff
@@ -303,94 +345,76 @@ class Server:
+            # Respond with the value if it exists
+            self.send_ok(msg, value=self.datastore.get(msg['key'], ''))
+        elif self.leader == 'FFFF':
             self.send_fail(msg)
+        else:
+            self.send_redirect(msg)
 
-    def handle_data(self, msg):
-        """Handles the data received from the socket accordingly"""
-        time = current_time()
-
-        # TODO: add a type for AppendEntries RPC responses
-        if msg['type'] == 'append_entries':
-            # Heartbeat
-            if not msg['entries']:
-                self.handle_heartbeat(msg)
-            # RPC from leader to replicate log entries
-            else:
-                self.handle_append_entries(msg)
-        # RPC to vote in an election
-        elif msg['type'] == 'request_vote':
-            self.handle_request_vote(msg)
-        # Vote received
-        elif msg['type'] == 'vote_granted' and self.state == State.CANDIDATE:
-            self.handle_received_vote(msg)
-        # Get request from client
-        elif msg['type'] == 'get':
-            self.handle_get(msg)
-        # Put request from client
-        elif msg['type'] == 'put':
-            self.handle_put(msg)
+    def apply(self, entries):
+        for e in entries:
+            if e['command'] == 'put':
+                self.datastore[e['key']] = e['value']
 
     def run(self):
         """Runs the main loop"""
         while True:
-            # Check for an election timeout
-            self.check_timeout()
+            time = current_time()
+            if self.state == State.LEADER:
+                span = time - self.prev_heartbeat
+                if span >= HEARTBEAT_FREQ:
+                    logging.info('Term {}: Sent Heartbeat'.format(self.current_term))
+                    self.broadcast_heartbeat()
+                    # Reset the heartbeat timer
+                    self.prev_heartbeat = time
 
-            ready = select.select([self.sock.sock], [], [], 0.1)[0]
-    
-            if self.sock.sock in ready:
-                msg = self.sock.recv()
+            if self.check_timeout():
+                # Election timeout met, so start election
+                self.transition_to_state(State.CANDIDATE)
+                self.send_request_vote()
+
+            if self.commit_idx > self.last_applied:
+                self.apply(self.log[self.last_applied:self.commit_idx + 1])
+                self.last_applied = self.commit_idx
+
+            msg = self.sock.recv(timeout=0.05)
+            if msg:
+                logging.debug('Replica ' + self.id + ' recieved ' +json.dumps(msg))
+                if 'term' in msg and self.current_term < msg['term']:
+                    logging.info('Reverting to FOLLOWER, Term {}.{} from {}'.format(msg['term'], msg['type'], msg['src']))
+                    self.transition_to_state(State.FOLLOWER, msg)
+                    self.reset_timeouts()
+                self.dispatch(msg['type'], msg)
```

Comment:
Can you explain what `dispatch()` does?

Reply:
Ooh nvm, I think I understand it now.

Reply:
Yeah I wrote an entire module to make that black magic happen